### ISSUE TITLE
python: NPA mis-counts class attributes (foreign-object writes over-count; tuple-unpack & multi-target binds under-count)

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#412]
## Summary

For Python, the NPA (number of public attributes) metric mis-counts class
attributes in three independent ways: it (a) over-counts by treating a write to
*any* object's attribute as an instance attribute, (b) under-counts
tuple/list-unpacking instance attributes, and (c) under-counts multi-target
class-level assignments.

## Location

- `src/metrics/npa.rs:1299` — `python_lhs_is_self_attribute` (foreign-object false positive)
- `src/metrics/npa.rs:1262-1285` — `python_self_attr_name_bytes` (requires the target be a single `Attribute`)
- `src/metrics/npa.rs:1184-1199` — `python_count_class_level_attrs` (one count per `=` statement)
- `src/metrics/npa.rs:736-738` — `total = class_level + self_attrs` (no cross-dedup; see Related)

## Evidence

All confirmed against the built `bca` CLI (`bca metrics -p f.py -O json`).

**(a) Foreign-object false positive.** `python_lhs_is_self_attribute` uses a
purely structural proxy — "an `Attribute` whose first child is an `Identifier`"
— and never reads the receiver text, even though `code` is available. So a write
to *any* object's attribute counts as an attribute of the class:

```python
class Service:
    def __init__(self, db, logger):
        self.name = "svc"      # real instance attribute
        db.connection = None   # foreign object — NOT an attribute of Service
        logger.level = "INFO"  # foreign object
```

Observed `npa.total = 3.0`; expected `1.0` (only `self.name`).

**(b) Tuple/list-unpacking under-count.** `python_self_attr_name_bytes` bails
unless `assignment.child(0)` is itself an `Attribute` (`npa.rs:1268`), so
unpacking targets (a `pattern_list`/`expression_list`) are skipped entirely:

```python
class C:
    def __init__(self):
        self.a, self.b = 1, 2   # target is a pattern_list, not Attribute
        self.c = 3
```

Observed `npa.total = 1.0` (only `self.c`); expected `3.0` (a, b, c).

**(c) Multi-target class-level under-count.** `python_count_class_level_attrs`
increments once per assignment statement carrying an `=`, ignoring how many
names it binds:

```python
class C:
    x = 1
    a = b = 3      # binds a and b
    p, q = 1, 2    # binds p and q
```

Observed class-level `npa.total = 3.0`; expected `5.0`. The Java sibling counts
each `VariableDeclarator`, so `int a, b;` yields 2 — Python diverges here.

## Expected Behavior

- `x.y = …` counts as a class attribute only when the receiver `x` is the
  method's `self`/`cls` alias, verified from source bytes — not for arbitrary receivers.
- `self.a, self.b = …` contributes one attribute per unpacked `self.<name>` target.
- A class-level assignment binding N names contributes N attributes.

## Actual Behavior

Foreign-object writes inflate the count; unpacking targets are dropped; and
multi-target class-level binds count as one.

## Impact

NPA is wrong for very common Python idioms: dependency wiring in `__init__`
(`self.db = db` alongside `db.x = …`), tuple-unpacking initialisation, and
grouped class-level constants. NPA feeds class-level reporting, so the error
propagates upward. Both the over-count (a) and the under-counts (b, c) can mask
or fabricate threshold breaches in `bca check`.

## Related

`total = class_level + self_attrs` (`npa.rs:738`) does no cross-dedup, so a
class-level default `x = 1` plus an instance `self.x = 2` counts `x` twice
(observed 2, arguably 1 — instance attribute shadows the class default). Lower
severity, same fix area.

---

## Resolution (fix/issue-412, commit a06a07fa)

- **(a) Foreign-object over-count — FIXED.** Receiver bytes are read
  and required to equal `self`/`cls` (literal byte match, the
  documented pragmatic option — not enclosing-function first-param
  resolution).
- **(b) Tuple/list-unpacking under-count — FIXED.** Each `self.<name>`
  element of a pattern_list/expression_list/tuple_pattern/list_pattern
  target is collected; mixed `self.a, x = …` counts only the self
  target.
- **(c) Multi-target class-level under-count — FIXED.** One attribute
  per bound name, following chained `a = b = 3` and unpacking
  `p, q = 1, 2`.
- **Dedup (sub-item 4) — IMPLEMENTED.** Class-level and instance passes
  share one name set; `x = 1` + `self.x = 2` counts `x` once. No large
  refactor was required, so it was not deferred.
- Edge: nested `self.f.g = 1` adds no class attribute (intentional).

Eight new anchored unit tests; each fix verified via test-via-revert.
No integration-snapshot drift (Python is not snapshotted in the output
submodule).

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

The Python NPA helpers in `src/metrics/npa.rs` approximate attribute detection
structurally and per-statement: `python_lhs_is_self_attribute` never reads the
receiver text, `python_self_attr_name_bytes` only handles a single `Attribute`
target, and `python_count_class_level_attrs` counts one per `=` statement.

### Implementation Steps

1. **Foreign-object false positive:** in `python_lhs_is_self_attribute` /
   `python_self_attr_name_bytes`, read the receiver identifier bytes and require
   them to equal `self`/`cls` — or, more robustly, the enclosing
   `function_definition`'s first parameter name. `code` is already threaded in.
2. **Unpacking under-count:** when `assignment.child(0)` is a `pattern_list` /
   `expression_list` / `tuple_pattern` / `list_pattern`, iterate its `Attribute`
   children and collect each `self.<name>` rather than bailing.
3. **Multi-target class-level under-count:** in `python_count_class_level_attrs`,
   count one per *bound name* — walk chained `a = b = 3` targets and the elements
   of a `p, q = …` target list.
4. **(Optional, Related)** cross-dedup class-level vs `self` attribute names
   before `total = class_level + self_attrs` (`npa.rs:738`).

### Tests to Add/Update

- Class with `self.x` plus foreign `db.conn = …` → `class_npa == 1`.
- `self.a, self.b = 1, 2` in `__init__` → `class_npa == 2` (+ any others).
- Class-level `a = b = 3` and `p, q = 1, 2` → counts each name.
- Use `check_metrics::<PythonParser>` with `assert_eq!` on `class_npa()` (anchor
  the snapshot per the project's snapshot-anchor rule).
- Mirror the new fixtures against the Java sibling to confirm parity.

### Verification

- [ ] `cargo test -p big-code-analysis` passes
- [ ] `cargo clippy -p big-code-analysis` clean
- [ ] `bca metrics` on the three repros in the issue returns 1 / 3 / 5 respectively

**Comment by dekobon** [comment]
Fixed on `fix/issue-412` (commit a06a07fa).

All three sub-bugs plus the optional dedup are addressed in-tree:

- **(a) Foreign-object over-count** — `python_self_attr_name_bytes`
  now reads the receiver bytes and requires them to equal `self` or
  `cls`. `db.connection = None` / `logger.level = "INFO"` no longer
  count. The Service example now reports `npa.total = 1.0`.
- **(b) Tuple/list-unpacking under-count** — a new
  `python_collect_self_attrs_from_target` iterates the elements of
  `pattern_list` / `expression_list` / `tuple_pattern` /
  `list_pattern` targets and collects each `self.<name>`. Mixed
  targets (`self.a, x = …`) count only the self attribute.
  `self.a, self.b = 1, 2; self.c = 3` now reports `3.0`.
- **(c) Multi-target class-level under-count** — class-level counting
  walks each bound name: chained `a = b = 3` (nested Assignment) and
  unpacking `p, q = 1, 2` each contribute two. `x=1; a=b=3; p,q=1,2`
  now reports `5.0`, mirroring Java's per-VariableDeclarator rule.
- **Dedup (sub-item 4)** — IMPLEMENTED (no large refactor needed).
  The class-level and instance passes share one name set, so a class
  default `x = 1` plus an instance `self.x = 2` counts `x` once.

Receiver-matching approach: literal `self`/`cls` byte comparison
(the documented pragmatic option), not enclosing-function first-param
resolution. Covers the vast majority of real code; a
non-conventionally-named first parameter would under-count, which is
preferable to the prior over-count of every foreign-object write.

Edge cases pinned with tests: nested `self.f.g = 1` adds no attribute
(receiver is an Attribute, not the self alias); classmethod `cls.x`
counts while a foreign `other.x` in the same body does not.

Eight new anchored unit tests added in `src/metrics/npa.rs`; each of
the three fixes was verified via test-via-revert. No integration
snapshot drift (the output submodule does not snapshot Python).
